lib/ansible/plugins/action/gather_facts.py

```diff
@@ -0,0 +1,86 @@
+# Copyright (c) 2017 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+import time
+
+from ansible.module_utils.common._collections_compat import MutableMapping
+
+from ansible import constants as C
+from ansible.plugins.action import ActionBase
+from ansible.utils.vars import combine_vars
+
+
+class ActionModule(ActionBase):
+
+    def _get_module_args(self, fact_module, task_vars):
+        mod_args = task_vars.get('ansible_facts_modules', {}).get(fact_module, {})
+        if isinstance(mod_args, MutableMapping):
+            mod_args.update(self._task.args.copy())
+        else:
+            mod_args = self._task.args.copy()
+
+        if fact_module != 'setup':
+            mod_args.pop('gather_subset', None)
```

Comment:
Should maybe add a warning here if gather_subset isn't `'all`? Alternatively, can we sanely tell if it was the gather_subset passed by config/play vs one set explicitly by the module args (so we're not effectively creating a new reserved module arg)?

Reply:
i expect that we phase out both 'setup' and gather_subset since this new feature lets us break out the 'huge' setup into ansible_min_facts, ansible_network_facts, etc 

But i'm fine with adding a warning here, if you really want it.